### PR TITLE
[CI] Add traceability on PR description

### LINKED ISSUE #9397: Add CI for traceability
## Proposed refactoring or deprecation

The goal would be to start validating the PR description.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

cc @borda @carmocca @akihironitta

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds a v0 CI for verifying the PR description. The goal is to progressively ask contributor to provide more details and tracking on the PR.

Fixes #9397

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/check-traceability.yml
Comment (kaushikb11) [inline .github/workflows/check-traceability.yml:36]: Could `os.getenv("PR_DESCRIPTION")` be None?

File: .github/workflows/check-traceability.yml
Comment (kaushikb11) [inline .github/workflows/check-traceability.yml:34]: regex doesn't seem to be used.

File: .github/workflows/check-traceability.yml
Comment (kaushikb11) [inline .github/workflows/check-traceability.yml:41]: Should we also check for `https://github.com/PyTorchLightning/pytorch-lightning/issues/xxxx` in the description?

File: .github/workflows/check-traceability.yml
Comment (SeanNaren) [inline .github/workflows/check-traceability.yml:39]:
```suggestion
          raise Exception("Please replace <issue_number> with a reference issue to track details of what this PR does!")
```

File: .github/workflows/check-traceability.yml
Comment (SeanNaren) [inline .github/workflows/check-traceability.yml:46]:
```suggestion
          raise Exception("Please link to an issue in the PR description to track details of what this PR!")
```

The wording of this I'm not entirely sure on, we should also give examples. I think we should also add the `related` tag?

Comment (kaushikb11) [reply .github/workflows/check-traceability.yml:46]: What if the PR is a minor docs update?

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:36]: No it can't

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:46]: The CI test is being tracked. minor docs update could also be linked quickly to a tiny issue.

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:41]: I think it is better to track for fixes, as it is used by Github to automate issue closing based on those words.

Comment (Borda) [reply .github/workflows/check-traceability.yml]: Missing space between fixes and #
How about other key words like fix, resolve, close, resolves, closes...

**Comment by Borda** [comment]
I would be curious if there is already available action to do this

File: .github/workflows/check-traceability.yml
Comment (awaelchli) [inline .github/workflows/check-traceability.yml:47]:
```suggestion
          raise Exception("Please link to an issue in the PR description to track details of what this PR does!")
```
the sentence is incomplete

Comment (awaelchli) [reply .github/workflows/check-traceability.yml:47]: isn't the exclamation mark a bit too aggressive?

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:7]: I'm not sure about validating this in CI. I don't think any minor changes should require an issue. And many external contributors open small PRs without linked issues for convenience.

Perhaps one option would be to allow setting "let's do it" on PRs to disable this check but then you wouldn't want to have to commit to re-trigger it. Would need to change the `on:` to something like message update

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:5]: This shouldn't apply as there's nothing to do after merged

```suggestion
```

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:18]: Name is already self-explainatory

```suggestion
    - uses: actions/setup-python@v2
```

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:26]: You can fetch the description with `gh pr view 2342 --json body --jq .body`

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:28]: Can we skip this if `gh pr view <Number> --json labels --jq 'any(.labels[].name; . == "let's do it!")'`

File: .github/workflows/check-traceability.yml
Comment (carmocca) [inline .github/workflows/check-traceability.yml:44]: Does this mean it needs to be written with no spaces?

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:7]: This is fine. I will enable this only for FB and core contributors. I think it is important we force traceability for all PRs.

Comment (carmocca) [reply .github/workflows/check-traceability.yml:7]: How would you do that? the workflow runs on all PRs

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:26]: I am using `| tr -d "\n\r" | tr -d ' '` to drop `space` and and newline as it won't be possible to add it to an env variable otherwise.

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:28]: I think every PR should be checked, possibly for us.

File: .github/workflows/check-traceability.yml
Comment (tchaton) [reply .github/workflows/check-traceability.yml:44]: No, spaces are being removed with `tr` tool.

**Review by carmocca — COMMENTED** [review]
This should run only on

```
on:
  pull_request:
    types: [edited, synchronize, opened, reopened]
    branches: [master]
```

Also, we could use https://github.com/hattan/verify-linked-issue-action

**Pull request closed without merging**
